### PR TITLE
Create Node API endpoints for Users

### LINKED ISSUE #633: Create Node API endpoints for Users
We need to create a variety of API endpoints for users to allow them to perform operations as data scientists and manage their node as compliance officers, administrators, and node owners. Here are a few commonalities between all endpoints:

- All endpoints should be supported in both HTTP and Websockets. Please follow the same organizational structure and workflow that was used in the static FL project for this.
- Successful returns should send back the user object itself. However, a few fields should never be sent with the response: **`hashed_password` and `salt` are not allowed in the return of any response of any kind.** Deceivingly, `private_key` **should** be returned with all requests because this key will be used to actually make API requests and authenticate tensor permissions.
- Successful returns should expand both the role field and the groups field to include the entire role object and array of group objects respectfully.
- Note that on these API routes, there is no `static` or `dynamic` prefix. This is because this table is shared between both workflows, and therefore the API endpoints should follow the same methodology.

### Create User - `POST /users`
**Request body:** `email`, `password`, `role`
**Description:** The default role of a new user will always be the lowest-level role: `User`. In fact, the `role` that is passed in the request body should be ignored unless the user making this request has the role of `can_create_users`. The `salt` field should be randomly generated for each user, combined with the plain-text password, and then hashed to generate the `hashed_password`. Please follow standard encryption protocols to generate and storing secure salts and hashed_passwords ([reference 1](https://crackstation.net/hashing-security.htm) and [reference 2](https://auth0.com/blog/adding-salt-to-hashing-a-better-way-to-store-passwords/)).

### Get User - `GET /users/{id}`
**Request body:** N/A
**Description:** Returns a single user. Only those with `can_triage_jobs` permissions can make this call.

### Get Users - `GET /users`
**Request body:** N/A
**Description:** Returns a list of users. Only those with `can_triage_jobs` permissions can make this call.

### Search Users - `POST /users/search`
**Request body:** `email`, `role`, `group`
**Description:** Returns a list of users, even if there are only 0 or 1 results. You can pass an `email`, `role`, or a `group`. If you pass multiple, it will perform an `AND` search (not an `OR`). Only those with `can_triage_jobs` permission can make this call.

### Change Email - `PUT /users/{id}/email`
**Request body:** `email`
**Description:** Changes the user's email address. You may only change your own email address. However, users with the `can_create_users` permission can also change the email addresses of other users.

### Change Password - `PUT /users/{id}/password`
**Request body:** `password`
**Description:** Changes the user's password. You may only change your own password. However, users with the `can_create_users` permission can also change the passwords of other users. Note that this should follow the same salting and hashing algorithm described in the **Create User** section above. Whenever a user changes their password, this should regenerate a new salt, instead of using the previous one.

### Change Role - `PUT /users/{id}/role`
**Request body:** `role`
**Description:** Changes the user's role. Only users with the `can_create_users` role can change another user's role. A few gotchas:
- The original Node Owner (also user with ID: `1`) cannot ever change their role to anything else.
- Only Owners can assign other users the role of `Owner`

### Change Groups - `PUT /users/{id}/groups`
**Request body:** `groups`
**Description:** Changes the user's group. Only users with the `can_create_users` role can change another user's groups. Note that this applies to both adding and removing a user from groups.

### Delete User - `DELETE /users/{id}`
**Request body:** N/A
**Description:** Deletes a user from the system. Only users with the `can_create_users` role can delete a user. However, a user can also delete themselves.

### Login - `POST /users/login`
**Request body:** `email`, `password`
**Description:** Logs a user into the system. Note that this does not need to create a session or anything like that, it merely returns the user object for that person if the credentials match.

### PR DESCRIPTION
**Pull request by Benardi**
## Description

We need to create a variety of API endpoints for users to allow them to perform operations as data scientists and manage their node as compliance officers, administrators, and node owners.

Fixes #633 

## Affected Dependencies
* Add package bcrypt

## How has this been tested?
* Unit tests for each endpoint (ws / http) including tests for:
   + bad request
   + missing resource
   + unauthorized operation
   + border cases
   + effect of requests on database

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: apps/node/src/app/main/routes/auth.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+from flask import current_app as app
+from syft.codes import RESPONSE_MSG
+import jwt
+
+from ..core.exceptions import MissingRequestKeyError, InvalidCredentialsError
+from ..users import User
+from ... import db
+
+
+def token_required(f):
```
Comment (cereallarceny) [inline apps/node/src/app/main/routes/auth.py:14]: Why can't we share this logic between here and `app/main/events/auth.py`?

File: apps/node/src/app/main/events/user_related.py [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+        response_body[RESPONSE_MSG.ERROR] = str(e)
+    except Exception as e:
+        response_body[RESPONSE_MSG.ERROR] = str(e)
+        logging.warning("User credentials are invalid", exc_info=e)
+
+    return dumps(response_body)
+
+
+@body_token_required
+def put_email_socket(current_user: User, message: dict) -> str:
```
Comment (cereallarceny) [inline apps/node/src/app/main/events/user_related.py:192]: Change all of the `put_` routes to describe what it's doing. This should be something like `change_email_socket`. Also make sure to do this for all the HTTP routes.

File: apps/node/src/app/main/users/user_ops.py [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+        raise AuthorizationError
+
+    user = User.query.get(user_id)
+    if user is None:
+        raise UserNotFoundError
+
+    return user
+
+
+def put_email(current_user, private_key, email, user_id):
```
Comment (cereallarceny) [inline apps/node/src/app/main/users/user_ops.py:155]: `change_email`, do this for all `put_` routes

**Review by cereallarceny — CHANGES_REQUESTED** [review]

File: apps/node/src/app/main/routes/auth.py (hunk 1 above)
Comment (Benardi) [reply apps/node/src/app/main/routes/auth.py:14]: The location  of the token changes from http to websocket so the function's logic is not quite the same

Comment (Benardi) [reply apps/node/src/app/main/routes/auth.py:14]: Also, the way the exceptions are handled also change, in one of them you gotta create a Response object, while in the other you simply return a object as a string

Comment (Benardi) [reply apps/node/src/app/main/routes/auth.py]: I've done it via [decorator factory](https://github.com/OpenMined/PyGrid/pull/708/files#diff-fe0006987fa850f55246e88f8dcb04ec)

image

**Review by cereallarceny — APPROVED** [review]

**Pull request merged**